include/seastar/testing/test_case.hh

```diff
@@ -54,3 +55,76 @@
             SEASTAR_TEST_CASE_WITHOUT_DECO,                         \
             SEASTAR_TEST_CASE_WITH_DECO),                           \
         __VA_ARGS__)
+
+namespace seastar::testing::internal {
+    template<typename T, typename = void>
+    struct has_setup : std::false_type {};
+    template<typename T>
+    struct has_setup<T, std::void_t<decltype(std::declval<T>().setup())>> : std::true_type {};
+    template<typename T>
+    seastar::future<> setup_conditional(T& t) {
+        if constexpr(has_setup<T>::value) {
+            return t.setup();
+        }
+        return seastar::make_ready_future();
+    }
+
+    template<typename T, typename = void>
+    struct has_teardown : std::false_type {};
+    template<typename T>
+    struct has_teardown<T, std::void_t<decltype(std::declval<T>().teardown())>> : std::true_type {};
+    template<typename T>
+    seastar::future<> teardown_conditional(T& t) {
+        if constexpr(has_setup<T>::value) {
+            return t.teardown();
+        }
+        return seastar::make_ready_future();
+    }
+}
+
+#define SEASTAR_FIXTURE_TEST_CASE_WITH_DECO(name, F, decorators)        \
+    struct name : public seastar::testing::seastar_test,                \
+                  private F {                                           \
+        using seastar::testing::seastar_test::seastar_test;             \
+        seastar::future<> run_test_case() override;                     \
+    private:                                                            \
+        seastar::future<> do_run_test_case() const;                     \
+    };                                                                  \
+    static const name name ## _instance(                                \
+        #name,                                                          \
+        __FILE__,                                                       \
+        __LINE__,                                                       \
+        decorators); /* NOLINT(cert-err58-cpp) */                       \
+    seastar::future<> name::run_test_case() {                           \
+        BOOST_TEST_CHECKPOINT('"' << #name << "\" fixture ctor");       \
+        BOOST_TEST_CHECKPOINT('"' << #name << "\" fixture setup");      \
+        /* seastar_test does not have  adefault ctor,  */               \
+        /* but setup_conditional needs to construct the fixture  */     \
+        /* to check if setup() is available, so cast it to F */         \
+        F& fixture = *this;                                             \
+        return testing::internal::setup_conditional(fixture).then([this] { \
+            BOOST_TEST_CHECKPOINT('"' << #name << "\" test entry");     \
+            return do_run_test_case();                                  \
+        }).finally([this] {                                             \
+            BOOST_TEST_CHECKPOINT('"' << #name << "\" fixture teardown"); \
```

Comment:
If I understand correctly, these fixtures are per-test-function fixtures, similar to pytest's "function-scoped" fixtures. They aren't shared by multiple test function. In that case, I honestly don't understand what's the point. In C++, you have RAII to do that - you can begin a test with a declaration `with_http_server h;` which will cause this type's constructor to run in the beginning at the test, and its destructor when the test ends. What does your fixture idea gain beyond this, or make more convenient?